cfp/37315 BGP - Auto Peer Discovery (#68)
cilium/design-cfps: cilium/CFP-37315-bgp-session-with-default-gateway.md

### Option 1:

When the peer IP is absent and there is exactly one peer, use the discovered IPv4 default gateway to create a single BGP session.

#### Pros

Simple implementation
Significantly reduces configuration overhead

#### Cons

Does not cover multi-homed networks or multiple peers.
Does not handle IPv6 or dual-stack scenarios.

### Option 2:

If both IPv4 and IPv6 default gateways are present and a single peer is defined without a peer IP, create two BGP sessions (IPv4 and IPv6). Provide configuration knobs to disable any undesired address families.

#### Pros

Adds flexibility for dual-stack environments.
Reduces the need for separate peer IP configurations for IPv4 and IPv6 by skipping the peer ip config

#### Cons

Deviates from the current "one bgp session per configured peer" design.
Still does not address multi-homed or multiple-peer scenarios.

## Future Milestones

_List things that this CFP will enable but that are out of scope for now. This can help understand the greater impact of a proposal without requiring to extend the scope of a CFP unnecessarily._

> ### Deferred Milestone 1

Author: should i put auto-discovery of multiple peers as a deferred milestone?

### Single Config File for All Nodes
Enable large networks to use a generic BGP configuration file without needing to specify unique peer IP addresses on a per-node basis.

## Non-Goals

* _List aspects which are specifically out of context for this CFP._

## Proposal

### Overview

At a high level, this proposal modifies how Cilium handles BGP peer IP configuration. If a user omits the peer IP for a BGP process, the Cilium agent attempts to retrieve the node’s default gateway from statedb (which stores local routing information). Cilium can then use this default gateway as the peer IP to establish the BGP session. If the default gateway is not found, the agent skips the neighbor and logs an error for it.

### BGP Cluster Config

If the user skips the peer ip for a bgp process and enables auto discovery, cilium agent should not return back an error when its creating new bgp sessions. It should first rely on statedb to get default gateway. If not found, it should skip the neighbor and log an error.
Every peer will have to specify the address family so that we fetch the default gateway for that address family

config proposal:

bgpInstances:
  - name: "65001"
    localASN: 65001

>     autoDiscovery: true

Author: ya i agree. it will provide flexibility per peer. but i'm curious about the use-case. What is the use-case to use different discovery mechanisms for each peer OR auto-discovery for one peer and static ip for another?

>     autoDiscovery: true

Reviewer: For example, using default gateway to peer with ToR and use static ip to peer with route reflector. Whether or not this is realistic, I'd prefer the option giving the maximum leverage. The instance-level knob enforces users to use default gateway for all peers.

>     autoDiscovery: true

Author: okay i was wondering if there is an existing use-case. 
but i agree with you. it provides flexibility for sure.

>     autoDiscovery: true

Author: updated the config and explanation

    peers:
    - name: "peer1 - 65000"
      peerASN: 65000
      afi: ipv4
      peerConfigRef:
        name: "cilium-peer"
    - name: "peer2 - 65000"
      peerASN: 65000
      afi: ipv6
      peerConfigRef:

>         name: "cilium-peer"

Author: fixed.

In multi-homed environments (i.e., multiple default gateways leading to different ASNs), this proposal does not address which gateway belongs to which peer ASN.

example multi-homed config:
bgpInstances:
  - name: "65001"
    localASN: 65001
    peers:
    - name: "65000"
      peerASN: 65000
      peerConfigRef:
        name: "cilium-peer"
    - name: "65011"
      peerASN: 65011
      peerConfigRef:
        name: "cilium-peer"

If multiple default gateways exist, we cannot reliably match them to different peer ASNs. One workaround is to use same local ASN on both the ToRs for bgp sessions with the k8s node, and bgp cluster config will have one peer configured.

proposed multi-homed config:
bgpInstances:
  - name: "65001"
    localASN: 65001
    peers:
    - name: "65000"
      peerASN: 65000
      afi: ipv4
      peerConfigRef:

>         name: "cilium-peer"

Reviewer: I think as long as the selection is consistent, it is ok to select only one of them and use it. For example, we can use metric to compare them. In my understanding, Linux doesn't allow us to create two default gateway with the same metric. So, it should be possible prefer the route with the lowest metric for example.

Since we are anyways subscribing to the netlink, we can also do failover when one of them disappears.

Reviewer: > are you saying that in multi-homing, we will create a bgp session with default gateway of lowest metric?

Yes, right. Then, when the lowest one disappears, we'll be notified through statedb (statedb watches netlink, so we don't need to deal with it directly). Then now we can select the one left on the routing table (the higher one). As a result, we failed over.

Author: okay got it. If you can send sample code or parts of cilium code that i can update to make this work, that would be helpful.

## Impacts / Key Questions

### Impact 1: Multi-Homing Scenario

In multi-homed environments, we will be breaking the design of one peer per bgp session because we will create two bgp sessions with one peer defined in the config. One of the workarounds could be to define two peers with same config and a unique name.

### Key Question

Is there a different way to accomplish this?

### Impact 2: Multiple BGP Instances

The proposal does not take into account multiple bgp instances so auto-discovery will work in that case.

### Key Question

What are use-cases of having multiple bgp instances? (bgp instances is a list in config)

### Impact 3: config knob for auto-discovery

config knob to enable auto-discovery will be crucial because that enables the users to understand the dependencies before making the changes. We can have documentation for auto-discover and provide the pre-requisites for enabling it.
Without the knob, the user may use auto-discovery without understanding the requirements for it.

### Key Question

> are we okay to add a knob for auto-discovery or should we enable auto-discovery by default if peerAddress is not specified?

Reviewer: I'd prefer an explicit knob. Also, that implicit knob only allows single type of the discovery mechanism. We may want to add different kind of discovery mechanism in the future.

Author: perfect! 100% agreed!

# CFP-37315

**SIG: SIG-BGP**

**Begin Design Discussion:** 2024-02-24

**Cilium Release:** X.XX

**Authors:** Naveen Achyuta <naveen.achyuta22@gmail.com>

## Summary

Cilium’s current BGP implementation requires specifying the peer IP address in the BGP cluster configuration. In large-scale environments with thousands of Kubernetes nodes, managing distinct BGP configuration files (one per node) becomes impractical.

This proposal offers a simpler alternative: rely on the default gateway of ToR switches to automatically create BGP sessions when no peer IP is provided in the BGP cluster configuration.

## Motivation

Large networks with thousands of Kubernetes nodes cannot feasibly manage the peer IP addresses of all ToR switches within numerous BGP cluster configurations. Allowing Cilium to automatically discover and use the  default gateway as the peer simplifies BGP session creation, reducing operational overhead and configuration complexity.

## Goals

### Auto-Discovery of Peer IPs

> Allow Cilium to establish BGP sessions by discovering the default gateway on the node, when no peer IP is explicitly provided in the BGP cluster configuration.

Reviewer: ```suggestion
Allow Cilium to establish BGP sessions with the nexthop IP address of the default gateway on the node.
```

### Single Config File for All Nodes

> Enable large networks to use a generic BGP configuration file without needing to specify unique peer IP addresses on a per-node basis.

Reviewer: ```suggestion
Enable large networks to use a generic BGP configuration file without needing to specify unique peer IP addresses on a per-rack basis.
```

## Non-Goals

> * _List aspects which are specifically out of context for this CFP._

Reviewer: ```suggestion
- Allow discovering the peer from arbitrary prefix ([ref](https://github.com/cilium/design-cfps/pull/68#issuecomment-2722443710))
```

## Proposal

### Overview

At a high level, this proposal modifies how Cilium handles BGP peer IP configuration. If a user omits the peer IP for a BGP process, the Cilium agent attempts to retrieve the node’s default gateway from statedb (which stores local routing information). Cilium can then use this default gateway as the peer IP to establish the BGP session. If the default gateway is not found, the agent skips the neighbor and logs an error for it.

### BGP Cluster Config

> If the user enables auto-discovery for a peer, cilium agent should use the discovery mode to get the peer ip. If default gateway mode is being used,
> tt should rely on statedb to get default gateway. If not found, it should skip the neighbor and log an error.
> Every peer will have to specify the address family under autoDiscovery config block so that we fetch the default gateway for that address family only.

Reviewer: ```suggestion
If the user enables auto-discovery for a peer, cilium agent should use the discovery mode to get the peer ip. If default gateway mode is being used,
it should rely on statedb to get default gateway. If not found, it should skip the neighbor and log an error.

Every peer will have to specify the address family under autoDiscovery config block so that we fetch the default gateway for that address family only.

If there are multiple default gateways, select the default gateway with the minimum metric. If it couldn't break the tie in this way, the reconciliation fails.
```

Without the knob, the user may use auto-discovery without understanding the requirements for it.

### Key Question

are we okay to add a knob for auto-discovery or should we enable auto-discovery by default if peerAddress is not specified?

### Option 1:

If the user wants to create bgp sessions without specifying peerAddress, the user will remove peerAddress field, enable auto-discovery in the config and add ipv4 and/or ipv6 peers explicitly with "afi" field

#### Pros

Adds flexibility to create ipv4 and/or ipv6 sessions.
auto-discovery field will help users to understand the requirements provided in the documentation before using it. (User can obviously enable it without knowing the requirements, but, we can error out with clear error messages)

#### Cons

For multi-homed environments, it deviates from the current "one bgp session per configured peer" design if we decide to use only one bgp peer for both sessions.

### Option 2:

do nothing

#### Pros

#### Cons

User will have to manage bgp cluster files for all the k8s nodes and updates to these files can be cumbersome

## Future Milestones

> _List things that this CFP will enable but that are out of scope for now. This can help understand the greater impact of a proposal without requiring to extend the scope of a CFP unnecessarily._

Reviewer: We can potentially mention about the `interface` field here if we decided to not implement it in the first iteration.

https://github.com/cilium/design-cfps/pull/68/files#r1993328948

Author: ya. I think we should add it to future milestones. It will be better to keep it simple and iterate over it when its needed. also, adding interface field should be easy to add in the future.

# CFP-37315

**SIG: SIG-BGP**

**Begin Design Discussion:** 2024-02-24

**Cilium Release:** X.XX

**Authors:** Naveen Achyuta <naveen.achyuta22@gmail.com>

## Summary

> Cilium’s current BGP implementation requires specifying the peer IP address in the BGP cluster configuration. In large-scale environments with thousands of Kubernetes nodes, managing distinct BGP configuration files (one per node) becomes impractical.

Author: i will fix that. it should be rack

### BGP Cluster Config

If the user skips the peer ip for a bgp process and enables auto discovery, cilium agent should not return back an error when its creating new bgp sessions. It should first rely on statedb to get default gateway. If not found, it should skip the neighbor and log an error.
Every peer will have to specify the address family so that we fetch the default gateway for that address family

config proposal:
```
bgpInstances:
  - name: "65001"
    localASN: 65001
    autoDiscovery: true
    peers:
    - name: "peer1 - 65000"
      peerASN: 65000
      afi: ipv4
      peerConfigRef:
        name: "cilium-peer"
    - name: "peer2 - 65000"
      peerASN: 65000
      afi: ipv6
      peerConfigRef:
        name: "cilium-peer"
```
In multi-homed environments (i.e., multiple default gateways leading to different ASNs), this proposal does not address which gateway belongs to which peer ASN.

> example multi-homed config:

Reviewer 2: Maybe we could add an optional config knob `interface` that could help in multi-homing scenario? E.g.:

```yaml
peers:
- name: peer0
  peerASN: 65000
  peerConfigRef:
    name: peer-config
  autoDiscovery:
    mode: default-gateway
    addressFamily: ipv4
    interface: eth0
- name: peer1
  peerASN: 65001
  peerConfigRef:
    name: peer-config
  autoDiscovery:
    mode: default-gateway
    addressFamily: ipv4
    interface: eth1
```

In this case, for `peer0` we would peer with the gateway based on the default route via `eth0`, for `peer1` we would peer with the gateway based on the default route via `eth1`.

Reviewer 3: Practically speaking, this might not be a big problem. Common scenario is server peering with two ToRs. So, it will be via different physical interfaces. I am okay to introduce optional `interface` field under `autoDiscovery` stanza.

Reviewer: The idea of the interface field sounds nice to me. With the BGP-based first hop redundancy scenario, this could help. Namely, the host is connected to two ToRs and have two default routes advertised from them with different metrics. In that case, on the FIB, there are two default routes with different interfaces.

Reviewer 4: I see where `interface` would be helpful, but I'd like to +1 the point about it being an optional setting.  Within the same cluster, and the same server SKU, our primary network interface shows up in Linux with a variety of names such as `enp193s0np0`.  The PCI slot used appears to be part of the name.

Author: yes i was about to mention that. Our initial proposal had interface as a way to solve multi-homing but i removed it because servers with pci slots can have different interface naming convention across different per nic model, per sku or per vendor. If a kubernetes cluster has different types of nic types or models, this will not work.

Author: i'm okay to add an optional interface config so that the user can decide on their own if they want to use it or not. 
but it still wont solve the problem for users that have different server vendors, thats why, we proposed our approach of using local-asn in the ToRs

Reviewer: Ah, indeed. "Predictable" interface name feature is always tricky. Users can still deal with that by adjusting [udev device renaming scheme](https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/7/html/networking_guide/sec-understanding_the_device_renaming_procedure#sec-Understanding_the_Device_Renaming_Procedure), but it also has another problem.

Yep, I think we can make it optional and even don't have to implement it from the beginning. It is still possible to add it without breaking the compatibility later. We can just take a note.

proposed multi-homed config:

```
bgpInstances:
  - name: "65001"
    localASN: 65001
    peers:
    - name: "65000"
      peerASN: 65000
      afi: ipv4
      peerConfigRef:
        name: "cilium-peer"
```

proposed config in the device (arista):
```
bgp listen range 10.2.0.0/24 peer-group SERVERS remote-as 65001
router bgp 65100
   router-id 10.1.1.1
   timers bgp 1 3
   neighbor SERVERS local-as 65000 no-prepend replace-as
```

### Routes from Statedb

cilium agent should rely on statedb to get the default gateway to create the bgp session. we can inject statedb into neighbor reconcillation struct and when we loop through neighbors, we can populate the peer ip by looking for default gateway of 0.0.0.0/0 or ::0/0 route in routes table of statedb.

## Impacts / Key Questions

### Impact 1: Multi-Homing Scenario

In multi-homed environments, we will be breaking the design of one peer per bgp session because we will create two bgp sessions with one peer defined in the config. One of the workarounds could be to define two peers with same config and a unique name.

### Key Question

Is there a different way to accomplish this?

### Impact 2: Multiple BGP Instances

> The proposal does not take into account multiple bgp instances so auto-discovery will work in that case.

Reviewer 2: I think what is proposed here may actually work in multi-instance setup as well, instance1 and instance2 could even peer with the same GW IP, with a different peer port in the peer config. Or instance1 can use auto-peering and instance2 static peering.

Author: got it. just curious can i know the use-case of having multiple instances of bgp?
